packages/passport/sdk/src/react/PassportContext.tsx

```diff
@@ -0,0 +1,242 @@
+});
+
+type PassportProviderProps = {
+  config: PassportModuleConfiguration;
+  children: React.ReactNode;
+};
+
+export function PassportProvider({ children, config }: PassportProviderProps) {
+  const [isLoggedIn, setLoggedIn] = useState(false);
+  const [isLoading, setIsLoading] = useState(false);
+  const [accounts, setAccounts] = useState<string[]>([]);
+  const [error, setError] = useState<Error | null>(null);
+  const passportInstance = new Passport(config);
+  const v = useMemo(() => ({
+    passportInstance,
+    isLoggedIn,
+    isLoading,
+    accounts,
+    error,
+    login: async () => {
+      try {
+        setError(null);
+        setIsLoading(true);
+        const provider = passportInstance.connectEvm();
+        const acc = await provider.request({ method: 'eth_requestAccounts' });
+        setLoggedIn(true);
+        setAccounts(acc);
+        return accounts;
+      } catch (e: any) {
+        setError(e);
+      } finally {
+        setIsLoading(false);
+      }
+      return [];
+    },
+    logout: async () => {
+      try {
+        setError(null);
+        setIsLoading(true);
+        await passportInstance.logout();
+        setLoggedIn(false);
+        setAccounts([]);
+      } catch (e: any) {
+        setError(e);
+      } finally {
+        setIsLoading(false);
+      }
+    },
+    loginWithoutWallet: async () => {
+      try {
+        setError(null);
+        setIsLoading(true);
+        const profile = await passportInstance.login();
+        setLoggedIn(true);
+        return profile;
+      } catch (e: any) {
+        setError(e);
+        return null;
+      } finally {
+        setIsLoading(false);
+      }
+    },
+    loginWithEthersjs: async () => {
+      try {
+        setError(null);
+        setIsLoading(true);
+        // eslint-disable-next-line new-cap
+        const provider = new Web3Provider(passportInstance.connectEvm());
+        const acc = await provider.send('eth_requestAccounts', []);
+        setLoggedIn(true);
+        return { accounts: acc, provider };
+      } catch (e: any) {
+        setError(e);
+        return { accounts: [], provider: null };
+      } finally {
+        setIsLoading(false);
+      }
+    },
+  }), [config, isLoggedIn]);
+
+  return (
+    <PassportContext.Provider value={v}>
+      {children}
+    </PassportContext.Provider>
+  );
+}
+
+export function usePassport() {
+  const c = useContext(PassportContext);
+  if (!c) {
+    throw new Error('usePassport must be used within a PassportProvider');
+  }
+  return c;
+}
+
+export function useIdToken() {
+  const { passportInstance, isLoading, isLoggedIn } = usePassport();
+  const [idToken, setIdToken] = useState<string | null>(null);
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<Error | null>(null);
+
+  const getIdToken = () => {
+    if (isLoggedIn) {
+      setLoading(true);
+      passportInstance
+        .getIdToken()
+        .then((t) => setIdToken(t || null))
```

Comment:
I think we'd want to handle `t` being undefined here in some way. What would cause this? And what would we expect to get back from `useIdToken` in this scenario?

Reply:
this is from calling oidcClient on this line: https://github.com/immutable/ts-immutable-sdk/blob/f4f6eee9ab99b61a2923f1625ada30d565098af6/packages/passport/sdk/src/authManager.ts#L525
So we don't really have control over.

Reply:
Hmm but in what scenario would this happen? If `isLoggedIn` is true and `getIdToken` returns undefined, should that set an error for example? Consider the result of useIdToken - it would be `{ idToken: null, error: null, loading: false }` - not helpful

Reply:
This should be equivalent to an internal error / bug scenario. I can set the error in this case. however, it complex things a bit more because now we need to consider if we need to clear up all the other fields.